nalgebra-lapack/src/colpiv_qr.rs

```diff
@@ -0,0 +1,777 @@
+    T: Scalar + ComplexField,
+    R: DimMin<C>,
+    C: Dim,
+{
+    /// Computes the orthonormal matrix `Q ∈ R^(m ⨯ n)` of this decomposition.
+    /// Note that this matrix has _economy_ dimensions, which means it is not
+    /// square unless `A` is square. It satisfies `Q^T Q = I`. Note further
+    /// that is is typically not necessary to compute `Q` explicitly. Rather,
+    /// check if some of the provided multiplication functions can help to
+    /// calculate the matrix products `Q B`, `B Q`, `Q^T B`, `B Q^T` more efficiently.
+    ///
+    /// This function allocates.
+    #[inline]
+    #[must_use]
+    pub fn q(&self) -> OMatrix<T, R, DimMinimum<R, C>>
+    where
+        DefaultAllocator: Allocator<R, <R as DimMin<C>>::Output>,
+        T: ColPivQrReal + Zero + ComplexField,
+    {
+        let (nrows, ncols) = self.qr.shape_generic();
+        let min_nrows_ncols = nrows.min(ncols);
+
+        if min_nrows_ncols.value() == 0 {
+            return OMatrix::zeros_generic(nrows, min_nrows_ncols);
+        }
+
+        let mut q = self
+            .qr
+            .generic_view((0, 0), (nrows, min_nrows_ncols))
+            .into_owned();
+
+        let mut info = 0;
+        let nrows = nrows.value() as i32;
+
+        let lwork = T::xorgqr_work_size(
+            nrows,
+            min_nrows_ncols.value() as i32,
+            self.tau.len() as i32,
+            q.as_mut_slice(),
+            nrows,
+            self.tau.as_slice(),
+            &mut info,
+        );
+
+        assert_eq!(check_lapack_info(info), Ok(()), "error in lapack backend");
+
+        let mut work = vec![T::zero(); lwork as usize];
+
+        T::xorgqr(
+            nrows,
+            min_nrows_ncols.value() as i32,
+            self.tau.len() as i32,
+            q.as_mut_slice(),
+            nrows,
+            self.tau.as_slice(),
+            &mut work,
+            lwork,
+            &mut info,
+        );
+
+        assert_eq!(check_lapack_info(info), Ok(()), "error in lapack backend");
+
+        q
+    }
+
+    /// Retrieves the upper trapezoidal submatrix `R` of this decomposition.
+    ///
+    /// This function allocates.
+    #[inline]
+    #[must_use]
+    pub fn r(&self) -> OMatrix<T, DimMinimum<R, C>, DimMinimum<R, C>>
+    where
+        DefaultAllocator: Allocator<DimMinimum<R, C>, DimMinimum<R, C>>,
+    {
+        let (nrows, ncols) = self.qr.shape_generic();
+        let d = nrows.min(ncols);
+        let m = self.qr.generic_view((0, 0), (d, d));
+        m.upper_triangle()
+    }
+}
+
+/// Utility trait to add a thin abstraction layer over lapack functionality for
+/// column pivoted QR decomposition.
+#[allow(missing_docs)]
+pub trait ColPivQrScalar: ComplexField + QRScalar {
+    /// routine for column pivoting QR decomposition using level 3 BLAS,
+    /// see https://www.netlib.org/lapack/lug/node42.html
+    /// or https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/geqp3.html
+    fn xgeqp3(
```

Comment:
I think that all functions of `ColPivQrScalar` must be `unsafe` because underlying lapack functions do not check the lengths of slices.

Reply:
You're right. Frankly, I think all the lapack trait functions in this crate must be unsafe. That is something I already have on my agenda for later, but I will definitely update the stuff inside my current PR.